compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts

```diff
@@ -251,16 +252,133 @@ export default function inferReferenceEffects(
             loc: eff.loc,
           });
         }
+        case 'ImmutableFunctionCall': {
+          // Handled below
+          break;
+        }
         default:
           assertExhaustive(
             eff,
             `Unexpected function effect kind \`${(eff as any).kind}\``,
           );
       }
     });
-  } else {
-    fn.effects = functionEffects;
+
+    if (functionEffects.length > 0) {
+      const error = new CompilerError();
+      let usedIdentifiers = new Set<IdentifierId>();
+      let names = new Map<IdentifierId, string | undefined>();
+
+      function visitFunction(fn: HIRFunction): void {
```

Comment:
If i'm understanding correctly, this seems like it's redoing part of the analysis of DCE to figure out what's used or not. That logic is pretty tricky, i'm wondering if makes sense for DCE record information on unused instructions or produce a Set of unused values? I don't love that either though, just a thought.